hathor/sysctl/runner.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2023 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import json
+from typing import TYPE_CHECKING, Any, Callable, Optional
+
+from hathor.sysctl.exception import (
+    SysctlEntryNotFound,
+    SysctlReadOnlyEntry,
+    SysctlRunnerException,
+    SysctlWriteOnlyEntry,
+)
+
+if TYPE_CHECKING:
+    from hathor.sysctl.sysctl import Sysctl
+
+
+class SysctlRunner:
+    """ Encapsulates the Sysctl to decouple it from the SyctlProtocol.
+    """
+
+    def __init__(self, root: 'Sysctl') -> None:
+        self.root = root
+
+    def run(self, line: str) -> bytes:
+        """Receives a string line, parses, interprets, acts over the Sysctl,
+        and returns an UTF-8 encoding data as feedback.
+        """
+        if not line:
+            raise SysctlRunnerException('line cannot be empty or None')
+
+        if line.startswith('!help'):
+            _, _, path = line.partition(' ')
+            return self.help(path)
+        elif line == '!backup':
+            return self.backup()
+
+        head, separator, tail = line.partition('=')
+        head = head.strip()
+        tail = tail.strip()
+        if separator == '=':
+            return self._set(head, tail)
+        else:
+            return self._get(head)
+
+    def _set(self, path: str, value_str: str) -> bytes:
```

Comment:
I don't like the fact that we are capturing all specific exceptions and turning them into `SysctlRunnerException` exceptions.

Reply:
Hmm... yeah we loose the original stack trace, however we gain the opportunity to deliver a better error message.

Reply:
I think that we just need better error messages for end users. Runners are used internally and they can fail raising a custom exception that can be easily captured by the caller. In this case, the protocol would capture these exceptions and generate friendly error messages.

I feel it's not a good design to convert all specific exceptions into a single type `SysctlRunnerException` in an internal class that might be used in different parts of the code (e.g., on the initialization of the node).

Reply:
Ok. Just to confirm, in this case I just need to drop the custom exception, right?

Reply:
I did this commit to remove the custom exceptions: 65ef8dbfb23fdaa5e6cd6248afe86b91c4169cf9

Reply:
I made this commit to treat the exceptions in the protocol: 09bde1c1744e81f3790f73f60e29d28579ff44cc

Reply:
Great! Thanks!